src/toastfile.rs

```diff
@@ -12,6 +13,77 @@ pub const DEFAULT_LOCATION: &str = "/scratch";
 // The default user for commands and files copied into the container
 pub const DEFAULT_USER: &str = "root";
 
+// This struct represents a path on the host and a corresponding path in the container.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct MappingPath {
+    pub host_path: PathBuf,
+    pub container_path: PathBuf,
+}
+
+impl Serialize for MappingPath {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(
+            format!(
+                "{}:{}",
+                self.host_path.to_string_lossy(),
+                self.container_path.to_string_lossy(),
+            )
+            .as_str(),
+        )
+    }
+}
+
+struct MappingPathVisitor;
+
+impl<'de> serde::de::Visitor<'de> for MappingPathVisitor {
+    type Value = MappingPath;
+
+    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+        formatter.write_str("cannot parse mapping path")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        if let Some((host_path, container_path)) = v.split_once(":") {
+            Ok(MappingPath {
+                host_path: host_path
+                    .to_owned()
+                    .parse()
+                    .map_err(|_| E::custom("illegal host path"))?,
+                container_path: container_path
+                    .to_owned()
+                    .parse()
+                    .map_err(|_| E::custom("illegal container path"))?,
+            })
+        } else {
+            Ok(MappingPath {
+                host_path: v
+                    .to_owned()
+                    .parse()
+                    .map_err(|_| E::custom("illegal host path"))?,
+                container_path: v
+                    .to_owned()
+                    .parse()
+                    .map_err(|_| E::custom("illegal container path"))?,
```

Comment:
Since the `host_path` and `container_path` are the same in this case, let's compute it once instead of twice. Maybe like this:

```rust
            let path: PathBuf = v.parse().map_err(|_| E::custom("Illegal host path."))?;

            Ok(MappingPath {
                host_path: path.clone(),
                container_path: path,
            })
```